### PR TITLE
adds an option to skip unversioned casks in outdated and upgrade command

### PR DESCRIPTION
**Pull request by bevanjkay**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Adds an additional flag `--skip-unversioned` for running brew outdated or upgrade. It skips casks that are given `version :latest`

Some documentation updates including within the terminal command outputs would also require updating.

Some of the areas I have added the flag as parameters may be superfluous at the moment.

### CODE REVIEW

**Comment by Logicer16** [comment]
See https://github.com/Homebrew/brew/issues/11513#issuecomment-871034005

File: Library/Homebrew/cmd/outdated.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/outdated.rb:41]:
```suggestion
      switch "--skip-unversioned",
```

**Review by MikeMcQuaid — COMMENTED** [review]
Makes sense to me but interested in any thoughts from @Homebrew/cask

**Comment by SMillerDev** [comment]
> Adds an additional flag --skip-unversioned for running brew outdated or upgrade. It skips casks that are given version :latest

Isn't that the default unless you pass `--greedy`?

**Comment by bevanjkay** [comment]
> > Adds an additional flag --skip-unversioned for running brew outdated or upgrade. It skips casks that are given version :latest
> 
> 
> 
> Isn't that the default unless you pass `--greedy`?

This is to be used in combination with greedy. I will need to document it properly. It updates casks with stanza `auto_updates true` but not `version :latest`

File: Library/Homebrew/cmd/outdated.rb
Comment (MikeMcQuaid) [reply Library/Homebrew/cmd/outdated.rb:41]: Also needs a `depends_on: "--greedy"

Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/outdated.rb:41]: what if this was `--skip-version-latest` or `--no-version-latest`?

Comment (bevanjkay) [reply Library/Homebrew/cmd/outdated.rb:41]: I think this is a good change --skip-version-latest looks like it would be more consistent with the other options.

**Comment by vitorgalvao** [comment]
> It updates casks with stanza `auto_updates true`

Which can mean a downgrade if the cask’s version is behind the app’s internal updater, which is not rare. Updating casks with `auto_updates true` is potentially more disruptive than updating ones with `version :latest`, which at worst will download the same thing.

So if adding a flag to skip `version :latest`, there should also be one to skip `auto_updates`.

That could be handled via an optional argument to `--greedy` or adding `--greedy-auto-updates` and `--greedy-latest` flags so the connection is clear.

**Comment by MikeMcQuaid** [comment]
> That could be handled via an optional argument to `--greedy` or adding `--greedy-auto-updates` and `--greedy-latest` flags so the connection is clear.

This kinda makes me wonder if it warrants either a default behaviour change or an environment variable to effectively set this flag/these flags all the time?

**Comment by vitorgalvao** [comment]
> This kinda makes me wonder if it warrants either a default behaviour change

Changing the default behaviour would, on every `upgrade`, either redownload already installed software or risk downgrading installed software or both.

The current default is the safest approach. `--greedy` should *not* be used. That’s part of the reason it was named that way, to make it clear there may be consequences.

> or an environment variable to effectively set this flag/these flags all the time?

Isn’t that `HOMEBREW_CASK_OPTS`? Doesn’t seem worth it to make new ones.

**Comment by bevanjkay** [comment]
> That could be handled via an optional argument to `--greedy` or adding `--greedy-auto-updates` and `--greedy-latest` flags so the connection is clear.

This looks to me like the best course of action. I will have a look at refactoring this PR when I get a chance. Thanks for your input @vitorgalvao

File: Library/Homebrew/cmd/outdated.rb
Comment (vitorgalvao) [inline Library/Homebrew/cmd/outdated.rb:39]:
```suggestion
             description: "Print outdated casks with `auto_updates true` or `version :latest`."
```

File: Library/Homebrew/cmd/outdated.rb
Comment (vitorgalvao) [inline Library/Homebrew/cmd/outdated.rb:45]:
```suggestion
             description: "Print outdated casks including those with `auto_updates true`."
```

File: Library/Homebrew/cask/cask.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/cask.rb:125]: Please use some parentheses here to make it clearer what you're intending.

**Comment by vitorgalvao** [comment]
And people complain of stale bots. They’re a good reminder to recheck issues; I had not realised the `do not merge` was already out.

@becanjkay If you’re happy with the current state of the PR, I’m good.

**Comment by bevanjkay** [comment]
@vitorgalvao As far as I am concerned it is good to go. My only concern would be that the parameters are being passed correctly through each method, as this is my first proper contribution to Homebrew/Brew

**Comment by vitorgalvao** [comment]
> My only concern would be ensuring that the parameters are being passed correctly through each method, as this is my first proper contribution to Homebrew/Brew

For that I’ll defer to other maintainers who usually merge to the core. But from an HBC interface standpoint, thumbs up from me.

**Review by miccal — APPROVED** [review]

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks again @bevanjkay!

**Pull request merged**
